autoload/neoformat.vim

```diff
@@ -75,74 +74,125 @@ function! s:neoformat(bang, user_input, start_line, end_line) abort
-            let lines_after = getbufline(bufnr('%'), a:end_line + 1, '$')
-            let lines_before = getbufline(bufnr('%'), 1, a:start_line - 1)
+  call neoformat#utils#log(a:job.stdin)
+  call neoformat#utils#log(cmd.exe)
 
-            let new_buffer = lines_before + stdout + lines_after
-            if new_buffer !=# original_buffer
+  call s:job_init(a:job)
 
-                call s:deletelines(len(new_buffer), line('$'))
+  if exists('*jobstart')
+       call s:job_run_jobstart(a:job)
+  else
+       call s:job_run_system(a:job)
+  endif
+endfunction
 
-                call setline(1, new_buffer)
 
-                call add(formatters_changed, cmd.name)
-                let endmsg = cmd.name . ' formatted buffer'
-            else
+function! s:job_run_system(job) abort
+    let cmd = a:job.cmd
+    let stdin = a:job.stdin
 
-                let endmsg = 'no change necessary with ' . cmd.name
-            endif
-            if !neoformat#utils#var('neoformat_run_all_formatters')
-                return neoformat#utils#msg(endmsg)
-            endif
-            call neoformat#utils#log('running next formatter')
-        else
-            call add(formatters_failed, cmd.name)
-            call neoformat#utils#log(v:shell_error)
-            call neoformat#utils#log('trying next formatter')
-        endif
-    endfor
-    if len(formatters_failed) > 0
-        call neoformat#utils#msg('formatters ' . join(formatters_failed, ", ") . ' failed to run')
+    if cmd.stdin
+        call neoformat#utils#log('using stdin')
+        let stdin_str = join(stdin, "\n")
+        let stdout = system(cmd.exe, stdin_str)
+    else
+        call neoformat#utils#log('using tmp file')
+        call writefile(stdin, cmd.tmp_file_path)
+        let stdout = system(cmd.exe)
     endif
-    if len(formatters_changed) > 0
-        call neoformat#utils#msg(join(formatters_changed, ", ") . ' formatted buffer')
-    elseif len(formatters_failed) == 0
-        call neoformat#utils#msg('no change necessary')
+
+    call s:job_stdout(a:job, [stdout])
+    call s:job_exit(a:job, v:shell_error)
+endfunction
+
+function! s:job_run_jobstart(job) abort
+    let cmd = a:job.cmd
+
+    let job_start_opts = {
+          \ 'on_exit': { job_id, exitcode, event -> s:job_exit(a:job, exitcode) },
+          \ 'on_stdout': { job_id, data, event -> s:job_stdout(a:job, data) },
+          \ 'stdout_buffered':v:true,
+          \ 'on_stderr': { job_id, data, event -> s:job_stderr(a:job, data) },
+          \ }
+
+    if cmd.stdin
+        call neoformat#utils#log('using stdin')
+    else
+        call neoformat#utils#log('using tmp file')
+        call writefile(a:job.stdin, cmd.tmp_file_path)
     endif
+
+    call neoformat#utils#log('doing jobstart ' . cmd.exe)
+    let id = jobstart(cmd.exe, job_start_opts)
+    call neoformat#utils#log('channel_id:' . id)
+    if cmd.stdin
+        call chansend(id, a:job.stdin)
```

Comment:
Since `chansend()` and `chanclose()` aren't available in old `nvim` versions, can we add a warning when a user is running `nvim` and the function doesn't exist?

Ideally at the top of `s:neoformat()`

something along the lines of

```viml

if has('nvim') && !exists("*chansend")
    neoformat#utils#warn("Your version of neovim is no longer supported. Consider upgrading.")
endif

Reply:
Also, does the latest vim support `chansend` and `chanclose`?

Reply:
It also has `job_start` intead of `jobstart` with different key names. I was planning to have a different function for vim8.

```diff
   if exists('*jobstart')
        call s:job_run_jobstart(a:job)
+  elseif exists('*job_start')
+       call s:job_run_job_start(a:job)
   else
        call s:job_run_system(a:job)
   endif

+function! s:job_run_job_start(job) abort
+    let cmd = a:job.cmd
+
+    let job_start_opts = {
+          \ 'in_mode': 'raw',
+          \ 'exit_cb': { job, exitcode -> s:job_exit(a:job, exitcode) },
+          \ 'out_cb': { ch, data -> s:job_stdout(a:job, [data]) },
+          \ 'out_mode': 'raw',
+          \ 'err_cb': { ch, data -> s:job_stderr(a:job, [data]) },
+          \ 'err_mode': 'raw',
+          \ }
+
+    if cmd.stdin
+        call neoformat#utils#log('using stdin')
+    else
+        call neoformat#utils#log('using tmp file')
+        call writefile(a:job.stdin, cmd.tmp_file_path)
+    endif
+
+    call neoformat#utils#log('doing jobstart ' . cmd.exe)
+    let id = job_start(cmd.exe, job_start_opts)
+    if job_status(id) == 'fail'
+      call neoformat#utils#log('failed to start')
+      " Documentation says:
+      "    If the job fails to start then job_status() on the returned
+      "    Job object results in "fail" and none of the callbacks will be
+      "    invoked.
+      call s:job_exit(a:job, -1)
+    endif
+    call neoformat#utils#log('channel_id:' . id)
+    let channel = job_getchannel(id)
+    if cmd.stdin
+        call ch_sendraw(channel, join(a:job.stdin, "\n"))
+    endif
+    call ch_close_in(channel)
+endfunction
```

But couldn't managed to make it work yet.

Reply:
Spent too much time on it, but, I made a docker-compose file of different VIM versions running the tests.

Turns out nvim 0.1 doesn't support Funcrefs at all, so I removed `Complete_cb` and `Log_cb` and now the old flow works for old nvim versions. 

I couldn't find an easy way to make nvim 0.2 compile (https://github.com/neovim/neovim/issues/8056), so I just added a check to support nvim-0.3+. This also avoids `jobsend` issue. It'll just use synchronous flow for old neovim.

What do you think?

Reply:
Sounds good.

I think falling back to synchronous is a good compromise.